src/Octoliner.h

```diff
@@ -1,102 +1,48 @@
-
-#ifndef DEFAULT_GPIOEXP_ADDR
-#define DEFAULT_GPIOEXP_ADDR 42
-#endif
-// DEFAULT_GPIOEXP_ADDR
+class Octoliner : public GpioExpander {
+public:
+    Octoliner(uint8_t i2cAddress = 42);
+    void begin();
+    void begin(TwoWire* wire);
+    void setSensitivity(uint8_t sense);
+    int16_t analogRead(uint8_t sensor);
+    void analogReadAll(int16_t* analogValues);
+    uint8_t digitalReadAll(void);
 
-// Pins definiion
-constexpr uint8_t LED_BRIGHTNESS_PIN = 9;
-constexpr uint8_t SENSITIVITY_PIN = 0;
+    float trackLine(void);
+    float trackLine(uint8_t pattern);
+    float trackLine(int16_t* analogValues);
 
-enum IOcommand {
-    WHO_AM_I //get UID
-    ,
-    RESET // reset
-    ,
-    CHANGE_I2C_ADDR // change I2C-address manually
-    ,
-    SAVE_I2C_ADDR // Save current address to flash
-    ,
-    PORT_MODE_INPUT // set pins to INPUT mode
-    ,
-    PORT_MODE_PULLUP // ... with pull-up internal resistor
-    ,
-    PORT_MODE_PULLDOWN // .. with pull-down internal resistor
-    ,
-    PORT_MODE_OUTPUT // set pins to OUTPUT mode
-    ,
-    DIGITAL_READ // read port (all of 8 pins)
-    ,
-    DIGITAL_WRITE_HIGH //
-    ,
-    DIGITAL_WRITE_LOW //
-    ,
-    ANALOG_WRITE //
-    ,
-    ANALOG_READ //
-    ,
-    PWM_FREQ // common pwm frequency to all pins
-    ,
-    ADC_SPEED,
-    SEND_MASTER_READED_UID,
-    CHANGE_I2C_ADDR_IF_UID_OK
-};
+    virtual uint8_t mapAnalogToPattern(int16_t* analogValues) const;
+    virtual float mapPatternToLine(uint8_t pattern) const;
 
-class Octoliner {
-public:
-    Octoliner();
-    Octoliner(uint8_t i2caddress);
-    void begin(uint8_t value);
-    void begin(void);
-    void digitalWrite(int pin, bool value);
-    void pinMode(int pin, uint8_t mode);
-    void analogWrite(int pin, uint8_t pulseWidth);
-    void pwmFreq(uint16_t freq);
-    void changeAddr(uint8_t newAddr);
-    void changeAddrWithUID(uint8_t newAddr);
-    void saveAddr();
-    void reset();
-    int digitalRead(int pin);
-    int analogRead(int pin);
-    uint32_t getUID();
-    int digitalReadPort();
-    void digitalWritePort(uint16_t value);
-    void pinModePort(uint16_t value, uint8_t mode);
-    void adcSpeed(uint8_t speed);
-    void setSensitivity(uint8_t sense);
-    void setBrightness(uint8_t brightness);
-    float mapLine(int binaryLine[8]);
+    // Deprecated mapLine
+    float mapLine(int* analogValues);
 
 private:
-    uint8_t _i2caddress;
-    float _value;
-    void writeCmdPin(IOcommand command, uint8_t pin, bool sendStop = true);
-    void writeCmdPin16Val(IOcommand command, uint8_t pin, uint16_t value, bool sendStop = true);
-    void writeCmd16BitData(IOcommand command, uint16_t data);
-    void writeCmd8BitData(IOcommand command, uint8_t data);
-    void writeCmd(IOcommand command, bool sendStop = true);
-    int read16Bit();
-    uint32_t read32bit();
+    const uint8_t _IRLedsPin = 9;
+    const uint8_t _sensePin = 0;
+    const uint8_t _sensorPinMap[8] = {
+        4, 5, 6, 8, 7, 3, 2, 1
+    };
```

Comment:
No need to keep these in RAM for every instance of the class. Better move them to C++ and make `constexpr`

Reply:
For single const - ok, but for array - it is bad idea (linker cant find const expression). I made "static const _sensorPinMap[8];"